From: Michal Kubecek <mkubecek@suse.cz>
Date: Fri, 10 Jan 2014 15:34:03 +0100
Subject: macvlan: introduce IFF_MACVLAN flag and helper function
Patch-mainline: v3.13-rc1
Git-commit: a6cc0cfa72e0b6d9f2c8fd858aacc32313c4f272 (partial)
References: bnc#846984

Introduce IFF_MACVLAN private flag to distinguish macvlan
devices and netif_is_macvlan() helper function to test it.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 drivers/net/macvlan.c     | 2 ++
 include/linux/netdevice.h | 5 +++++
 include/uapi/linux/if.h   | 1 +
 3 files changed, 8 insertions(+)

diff --git a/drivers/net/macvlan.c b/drivers/net/macvlan.c
index 0924e51b..cf30dee 100644
--- a/drivers/net/macvlan.c
+++ b/drivers/net/macvlan.c
@@ -844,6 +844,8 @@ int macvlan_common_newlink(struct net *src_net, struct net_device *dev,
 	if (err < 0)
 		goto upper_dev_unlink;
 
+	dev->priv_flags |= IFF_MACVLAN;
+
 	list_add_tail_rcu(&vlan->list, &port->vlans);
 	netif_stacked_transfer_operstate(lowerdev, dev);
 
diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 25f5d2d..91008cc 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -2945,6 +2945,11 @@ static inline void netif_set_gso_max_size(struct net_device *dev,
 	dev->gso_max_size = size;
 }
 
+static inline bool netif_is_macvlan(struct net_device *dev)
+{
+	return dev->priv_flags & IFF_MACVLAN;
+}
+
 static inline bool netif_is_bond_master(struct net_device *dev)
 {
 	return dev->flags & IFF_MASTER && dev->priv_flags & IFF_BONDING;
diff --git a/include/uapi/linux/if.h b/include/uapi/linux/if.h
index 1ec407b..d758163 100644
--- a/include/uapi/linux/if.h
+++ b/include/uapi/linux/if.h
@@ -83,6 +83,7 @@
 #define IFF_SUPP_NOFCS	0x80000		/* device supports sending custom FCS */
 #define IFF_LIVE_ADDR_CHANGE 0x100000	/* device supports hardware address
 					 * change when it's running */
+#define IFF_MACVLAN 0x200000		/* Macvlan device */
 
 
 #define IF_GET_IFACE	0x0001		/* for querying only */
-- 
1.8.1.4

